File: tools/octobeat/octobeat/models/metadata.py
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class CatalogMetadata(BaseModel):
    """
    Catalogue metadata describing a resource.

    This model matches the metadata.json/catalog.json documents
    consumed by OctoBeat (packages/library MetadataSchema).
    """

    model_config = ConfigDict(
        extra="forbid",
        populate_by_name=True,
    )

    id: str

    title: str

    artist: str

    album: str | None = None

    year: int | None = None

    genres: list[str] = Field(
        default_factory=list,
    )

    bpm: float = Field(gt=0.0)

    duration: float = Field(ge=0.0)

    difficulty: int | None = None

    tags: list[str] = Field(
        default_factory=list,
    )

    timeSignature: str | None = None

    timing: TimingProvenance | None = None

    resources: ResourceRefs


class TimingProvenance(BaseModel):
    """
    Where the timing information came from and how reliable it is.
    """

    model_config = ConfigDict(
        extra="forbid",
    )

    source: str

    confidence: float = Field(ge=0.0, le=1.0)


class ResourceRefs(BaseModel):
    """
    Relative media paths inside a resource directory.
    """

    model_config = ConfigDict(
        extra="forbid",
    )

    audio: str

    lyrics: str | None = None

File: packages/library/src/metadata.ts
import { z } from "zod";

export const ResourcesSchema = z.object({
    audio: z.string(),

    video: z
        .string()
        .nullish(),
});

export const TimingProvenanceSchema =
    z.object({
        source: z.string(),

        confidence: z.number(),
    });

export const MetadataSchema =
    z.object({
        id: z.string(),

        title: z.string(),

        artist: z.string(),

        album: z
            .string()
            .nullish(),

        year: z
            .number()
            .int()
            .nullish(),

        genres: z.array(
            z.string(),
        ),

        bpm: z.number(),

        duration: z.number(),

        difficulty: z
            .number()
            .int()
            .nullish(),

        tags: z.array(
            z.string(),
        ),

        timeSignature: z
            .string()
            .nullish(),

        timing: TimingProvenanceSchema
            .nullish(),

        youtube: z
            .string()
            .nullish(),

        resources:
            ResourcesSchema,
    });

export type Metadata =
    z.infer<
        typeof MetadataSchema
    >;

export function parseMetadata(
    json: unknown,
): Metadata {
    return MetadataSchema.parse(
        json,
    );
}